src/NuGet.Core/NuGet.Common/PathUtil/DirectoryUtility.cs

```diff
@@ -0,0 +1,50 @@
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+
+namespace NuGet.Common
+{
+    /// <summary>
+    /// Directory operation helpers.
+    /// </summary>
+    public static class DirectoryUtility
+    {
+        private static object s_sharedFolderCreationLock = new object();
+
+        /// <summary>
+        /// Creates all directories and subdirectories in the specified path unless they already exist.
+        /// New directories can be read and written by all users.
+        /// </summary>
+        public static void CreateSharedDirectory(string path)
+        {
+            if (RuntimeEnvironmentHelper.IsWindows)
+            {
+                Directory.CreateDirectory(path);
+            }
+            else
+            {
+                lock (s_sharedFolderCreationLock)
+                {
+                    path = Path.GetFullPath(path);
+                    var root = Path.GetPathRoot(path);
+                    var sepPos = root.Length - 1;
+                    do
+                    {
+                        sepPos = path.IndexOf(Path.DirectorySeparatorChar, sepPos + 1);
+                        var currentPath = sepPos == -1 ? path : path.Substring(0, sepPos);
+                        if (!Directory.Exists(currentPath))
+                        {
+                            Directory.CreateDirectory(currentPath);
+                            chmod(currentPath, UGO_RWX);
+                        }
+                    } while (sepPos != -1);
```

Comment:
What is this logic trying to do?  I'm not sure it really accomplishes the goal.  For example, the lock is only process-wide, so if the goal here is to address cross-process, you still have a race condition, e.g. after the check for Directory.Exists and before the chmod, another process/user could create the directory.

My recommendation:
- Check if the directory exists.  If it does, you're done.
- If it doesn't, create a temporary directory at a temporary location and change the permissions on it, then try to move that to the desired location.
Essentially this is an interlocked compare exchange, but for directories.